### PR TITLE
Parse tuple

### PR DESCRIPTION
**Pull request by MortalFlesh**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -166,6 +168,28 @@ GET http://host/endpoint/?field[lte]=value
...
 ```
+- `Tuples` are not allowed in `IN` filter
+
+## `Tuples` in filters
+`Tuples`
+- are important in filters if you have some values, which **must** be send together
+- are composed from two or more values (_`Tuple` of one value is just a value_)
+- items should be in `(` `)` and separated by `,`
+- it is advised NOT to use a `space` between values because of the `URL` specific behaviour
+- for more information about `Tuples` see https://github.com/MortalFlesh/MFCollectionsPHP#immutabletuple
```
Comment (legendik) [inline README.md]: The link should be linkable :)

**Review by legendik — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (MortalFlesh) [reply README.md]: it is .. github does it autmatically with links..
https://github.com/MortalFlesh/api-filter/tree/feature/tuple#tuples-in-filters

Comment (legendik) [reply README.md]: Cool, didn't know that!

**Review by legendik — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -166,6 +168,28 @@ GET http://host/endpoint/?field[lte]=value
 ```http request
 GET http://host/endpoint/?type[in][]=one&type[in][]=two
 ```
+- `Tuples` are not allowed in `IN` filter
```
Comment (OndraM) [inline README.md:171]: I prefer using backticks for code-related excerpts, like method names, variables etc. Not for term emphasis - what about italics? _Tuples_ etc.

Same as below - why backticks around URL, space and other terms? For emphasis use either _italics_ or __bold__.

File: README.md [hunk 3]
```diff
@@ -166,6 +168,28 @@ GET http://host/endpoint/?field[lte]=value
...
+
+## `Tuples` in filters
+`Tuples`
+- are important in filters if you have some values, which **must** be send together
```
Comment (OndraM) [inline README.md]: be sent

File: README.md [hunk 4]
```diff
@@ -166,6 +168,28 @@ GET http://host/endpoint/?field[lte]=value
...
+- are composed from two or more values (_`Tuple` of one value is just a value_)
```
Comment (OndraM) [inline README.md]: "are composed of two ..." or "consist of two..."

File: README.md [hunk 5]
```diff
@@ -166,6 +168,28 @@ GET http://host/endpoint/?field[lte]=value
...
+- items should be in `(` `)` and separated by `,`
+- it is advised NOT to use a `space` between values because of the `URL` specific behaviour
```
Comment (OndraM) [inline README.md]: behavior

File: README.md [hunk 6]
```diff
@@ -166,6 +168,28 @@ GET http://host/endpoint/?field[lte]=value
...
+- for more information about `Tuples` see https://github.com/MortalFlesh/MFCollectionsPHP#immutabletuple 
+
+### Column with `Tuple`
+Columns declared by `Tuple` behaves the same as a single value but its value must be a `Tuple` as well.
+
+### Values with `Tuple`
+Values in the `Tuple` must have same number of items as is number of columns.
```
Comment (OndraM) [inline README.md]: the same

Comment (OndraM) [reply README.md]: the number of columns

File: README.md [hunk 7]
```diff
@@ -166,6 +168,28 @@ GET http://host/endpoint/?field[lte]=value
...
+
+### Usage
+```http request
+GET http://host/endpoint/?(first,second)[eq]=(one,two) 
+```
+☝ means that you have two columns `first` and `second` and they must be send together.
```
Comment (OndraM) [inline README.md]: be sent

File: src/Service/QueryParametersParser.php [hunk 8]
```diff
@@ -17,23 +19,48 @@ public function parse(array $queryParameters): FiltersInterface
...
-                $values = is_array($values)
-                    ? $values
-                    : ['eq' => $values];
+                $columns = $this->parseColumns($column);
+                $columnsCount = count($columns);
 
-                foreach ($values as $filter => $value) {
-                    yield Tuple::of($column, $filter, new Value($value));
+                foreach ($this->normalizeFilters($values) as $filter => $value) {
+                    Assertion::false($columnsCount > 1 && $filter === 'in', 'Tuples are not allowed in IN filter.');
```
Comment (OndraM) [inline src/Service/QueryParametersParser.php]: What about moving the assertion to a private method more with readable name - eg. `assertTupleIsAllowed()`?

Comment (OndraM) [inline src/Service/QueryParametersParser.php]: This assertion could be moved to private method more with readable name - eg. `assertTupleIsAllowed()`

**Review by OndraM — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (MortalFlesh) [reply README.md:171]: Well I see a `Tuple` same as an `array` or any other type, so it is `code-related` for me.

For others (`space`, `URL`, ...) it just feels right to be in backticks. ¯\\_(ツ)_/¯

**Comment by danger-lmc** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  There are some fixup/squash commits that needs to be applied before merge.
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Pull request merged**